rstudio/shiny
# Automatically load 'helpers' in R/ directory at runtime

Sources files ending in `.r` (case-insensitive) found in the `R/` directory (case-sensitive) when starting the app. If using `app.R`, the helpers are sourced before `app.R`. If using `server.R`, the `global.R` file is source first, then the helpers, then your `server.R` and `ui.R` files (when needed).

### Environments

#### With `server.R` and `ui.R`:

```
+ global env (global.R)
|
+-+ helpers env (R/ files)
| |
| +-- ui.R env
| |
| +-- server.R env
```

The only change here is the introduction of a new environment in between the child ui and server environments into which the helpers are loaded.

#### With `app.R`

```
+ global env
|
+-+ helpers env (R/ files)
| |
| +-- app.R
```

### Breaking changes

 - We automatically source all scripts in the top level of R/ when an app is run

### Testing Notes

1. Try running the shiny example 168 against Shiny installed from CRAN. I have this app [here](https://github.com/rstudio/shiny-examples/compare/jeff/helpers2) (not merged, since it doesn't work against Shiny on master). You'll see that it fails with an error like:

```
ERROR: could not find function "counterButton"
```

2. Install this branch: `remotes::install_github("jeff/feature/helpers")`
3. Restart R
4. Rerun example app 168 and confirm that the app now runs. You should see a button that you can click on to increment a counter. If that works, that demonstrates that the R file in the R/ directory was automatically loaded.
